### PR TITLE
Fix finding references to active pattern cases

### LINKED ISSUE #14206: `FindBackgroundReferencesInFile` on Active Pattern Case finds all other Cases too
"Find All References" in Visual Studio (and Ionide) on an Active Pattern Case sometimes not just finds current case, but all other cases too:  
VS-FindAllReferences-Project-EvenOdd  
-> "Find All References" on `Even` -- but finds `Odd` too

But this isn't always the case:  
VS-FindAllReferences-Script-Even  
Here VS correctly finds just `Even` occurrences.

Difference: First one is a fs file inside a Project, second one is a Script file (even one not saved on disk)

<br/>

Also an issue for Renaming:  
Instead of just renaming `Even`, `Odd` gets renamed at same time (to same name)

<br/>

---------------------
Side Note:  
Images above show a presentation error too:  
`let |Even|Odd|(|Even|Odd|) v =` instead of `let (|Even|Odd|) v =`  
Link location (click on result) works fine

---------------------

<br/>

The difference in behaviour can be explained by looking at FindUsagesService:  
https://github.com/dotnet/fsharp/blob/d13df68121219e88f50632a35dda872afca23eb7/vsintegration/src/FSharp.Editor/Navigation/FindUsagesService.fs#L103-L126  
-> different lookup methods depending on local to current doc or public:
* current doc: `checkFileResults.GetUsesOfSymbolInFile`
* otherwise: `SymbolHelpers.getSymbolUsesInProjects` which boils down to `checker.FindBackgroundReferencesInFile` ([actual call](https://github.com/dotnet/fsharp/blob/d13df68121219e88f50632a35dda872afca23eb7/vsintegration/src/FSharp.Editor/LanguageService/WorkspaceExtensions.fs#L186))

(Ionide uses same functions for same scopes -> explains same behavior as in VS)

<br/>

So it seems `FSharpCheckFileResults.GetUsesOfSymbolInFile` and `FSharpChecker.FindBackgroundReferencesInFile` return different results for Active Pattern Cases -- with `FSharpChecker.FindBackgroundReferencesInFile` being wrong

<br/>
<br/>

**Repro steps**

([Test Script](https://gist.github.com/Booksbaum/275f6af13410812ba97a0647b9d2408f#file-test-fsx))

Tests find usages on last `Even` in:
```fsharp
let (|Even|Odd|) v = 
  if v % 2 = 0 then Even else Odd
match 2 with
| Even -> ()
| Odd -> ()
```
<br/>

> `git clone https://gist.github.com/275f6af13410812ba97a0647b9d2408f.git test`
>
> `cd test`
>
> `dotnet fsi --nologo --exec .\Test.fsx`

Output:
```
# Project
* uses: Ok
* refs: Error
    Expected 3, but was 6: [|...\src\Project.fs (2,6--2,10);
  ...\src\Project.fs (2,11--2,14);
  ...\src\Project.fs (3,20--3,24);
  ...\src\Project.fs (3,30--3,33);
  ...\src\Project.fs (5,2--5,6);
  ...\src\Project.fs (6,2--6,5)|]
# Script
* uses: Ok
* refs: Error
    Expected 3, but was 6: [|...\src\Script.fsx (2,6--2,10);
  ...\src\Script.fsx (2,11--2,14);
  ...\src\Script.fsx (3,20--3,24);
  ...\src\Script.fsx (3,30--3,33);
  ...\src\Script.fsx (5,2--5,6);
  ...\src\Script.fsx (6,2--6,5)|]
# InMemory
* uses: Ok
* refs: Error
    Exception: System.AggregateException: [...]
    // Expected, because no existing file
```

* `uses`: `GetUsesOfSymbolInFile`
* `refs`: `FindBackgroundReferencesInFile`
* `Ok` when 3 occurrences found, otherwise `Error`

-> tests prove: Different results for `GetUsesOfSymbolInFile` and `FindBackgroundReferencesInFile` and wrong results for `FindBackgroundReferencesInFile`

**Expected behavior**

`FindBackgroundReferencesInFile` should only return locations of search Active Pattern Case.  

**Actual behavior**

`FindBackgroundReferencesInFile` returns locations of other Cases too

**Known workarounds**

Use `GetUsesOfSymbolInFile` instead of `FindBackgroundReferencesInFile`

**Related information**

* Win 11 x64
* `dotnet --version`: `6.0.402`, `7.0.100-rc.2.22477.23`
* FCS: `41.0.5`, `41.0.6`, `42.7.100-preview.22473.1`
* Microsoft Visual Studio Community 2022 (64-bit) - Current Version 17.3.6

### PR DESCRIPTION
**Pull request by 0101**
Fixes #14206.

When we deal with `Item.ActivePatternResult` - which comes from the definition of the pattern, we don't have access to enclosing entity. Therefore we cannot properly uniquely identify the definition in the item key store. We can almost do it with a combination of all the pattern cases + filename, which we can find from range. Therefore this solution won't work correctly if someone will have multiple active patterns with the same cases in the same order defined in the same file. But I think we can live with that. (And it was also the case before this fix.)

Another unsolved issue is that we don't find active patterns in signature files, because there they're represented as `Item.Value` from which we can't tell that it's an active patter at all. But combination of using active patterns and signature files should be pretty rare so we can probably live with that as well.

### CODE REVIEW

File: tests/FSharp.Compiler.ComponentTests/FSharpChecker/FindReferences.fs [hunk 1]
```diff
@@ -301,3 +301,104 @@ let x = MyType()
...
+            ])
+        }
+
+    [<Fact>]
+    let ``We find active patterns in other files when there are signature files`` () =
+
+        SyntheticProject.Create(
+            { sourceFile "First" [] with
+                Source = "let (|Even|Odd|) v = if v % 2 = 0 then Even else Odd"
+                SignatureFile = AutoGenerated },
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/FSharpChecker/FindReferences.fs:391]: So do I understand correctly that renaming an active pattern case will still not lead to its renaming in the signature file (and hence there'll be a compiler error)?

File: tests/FSharp.Compiler.ComponentTests/FSharpChecker/FindReferences.fs [hunk 2]
```diff
@@ -301,3 +301,104 @@ let x = MyType()
...
+
+    let (|Even|Odd|) v =
+        if v % 2 = 0 then Even else Odd
+    match 2 with
+    | Even -> ()
+    | Odd -> ()
+
+module Two =
+
+    let (|Even|Steven|) v =
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/FSharpChecker/FindReferences.fs:344]: How about Odd Steven?

**Review by psfinaki — APPROVED** [review]
Judging by the PR description, I'd rather call this as "improving finding references" than "fixing finding references" - but that's on you. :)

File: tests/FSharp.Compiler.ComponentTests/FSharpChecker/FindReferences.fs (hunk 1 above)
Comment (0101) [reply tests/FSharp.Compiler.ComponentTests/FSharpChecker/FindReferences.fs:391]: Yes. Signature files ruin everything 😭 

To make this work would probably require significant changes.

Comment (0101) [reply tests/FSharp.Compiler.ComponentTests/FSharpChecker/FindReferences.fs:391]: Yeah. Created https://github.com/dotnet/fsharp/issues/14969 and added a test here documenting the broken behavior.

File: tests/FSharp.Compiler.ComponentTests/FSharpChecker/FindReferences.fs (hunk 2 above)
Comment (0101) [reply tests/FSharp.Compiler.ComponentTests/FSharpChecker/FindReferences.fs:344]: No.

**Review by T-Gro — APPROVED** [review]

**Pull request merged**